metadata/src/from_into/v14.rs

```diff
@@ -178,17 +226,190 @@ fn v14_to_v15(metadata: v14::RuntimeMetadataV14) -> v15::RuntimeMetadataV15 {
                 frame_metadata::v15::SignedExtensionMetadata {
                     identifier: ext.identifier,
                     ty: ext.ty,
                     additional_signed: ext.additional_signed,
                 }
-            }).collect()
+            }).collect(),
+            address_ty: extrinsic_parts.address.into(),
+            call_ty: extrinsic_parts.call.into(),
+            signature_ty: extrinsic_parts.signature.into(),
+            extra_ty: extrinsic_parts.extra.into(),
         },
         ty: metadata.ty,
         apis: Default::default(),
+        outer_enums,
+        custom: v15::CustomMetadata {
+            map: Default::default(),
+        },
+    }
+}
+
+/// The type IDs extracted from the metadata that represent the
+/// generic type parameters passed to the `UncheckedExtrinsic` from
+/// the substrate-based chain.
+struct ExtrinsicPartTypeIds {
+    address: u32,
+    call: u32,
+    signature: u32,
+    extra: u32,
+}
+
+impl ExtrinsicPartTypeIds {
+    /// Extract the generic type parameters IDs from the extrinsic type.
+    fn new(metadata: &v14::RuntimeMetadataV14) -> Result<Self, String> {
+        const ADDRESS: &str = "Address";
+        const CALL: &str = "Call";
+        const SIGNATURE: &str = "Signature";
+        const EXTRA: &str = "Extra";
+
+        let extrinsic_id = metadata.extrinsic.ty.id;
+        let Some(extrinsic_ty) = metadata.types.resolve(extrinsic_id) else {
+            return Err("Missing extrinsic type".into())
+        };
+
+        let params: HashMap<_, _> = extrinsic_ty
+            .type_params
+            .iter()
+            .map(|ty_param| {
+                let Some(ty) = ty_param.ty else {
+                    return Err("Missing type param type from extrinsic".to_string());
+                };
+
+                Ok((ty_param.name.as_str(), ty.id))
+            })
+            .collect::<Result<_, _>>()?;
+
+        let Some(address) = params.get(ADDRESS) else {
+            return Err("Missing address type from extrinsic".into());
+        };
+        let Some(call) = params.get(CALL) else {
+            return Err("Missing call type from extrinsic".into());
+        };
+        let Some(signature) = params.get(SIGNATURE) else {
+            return Err("Missing signature type from extrinsic".into());
+        };
+        let Some(extra) = params.get(EXTRA) else {
+            return Err("Missing extra type from extrinsic".into());
+        };
+
+        Ok(ExtrinsicPartTypeIds {
+            address: *address,
+            call: *call,
+            signature: *signature,
+            extra: *extra,
+        })
     }
 }
+
+fn generate_outer_enums(
+    metadata: &mut v14::RuntimeMetadataV14,
+) -> v15::OuterEnums<scale_info::form::PortableForm> {
+    let call_enum = metadata
+        .types
+        .types
+        .iter()
+        .find(|ty| {
+            let Some(ident) = ty.ty.path.ident() else { return false };
```

Comment:
I don't follow the `return false` thingy here, if `an Ident` is missing what does that mean?